homeassistant/components/unifiprotect/utils.py

```diff
@@ -19,3 +36,88 @@ def get_nested_attr(obj: Any, attr: str) -> Any:
         value = value.value
 
     return value
+
+
+async def profile_ws_messages(
+    hass: HomeAssistant,
+    protect: ProtectApiClient,
+    seconds: int,
+    device_entry: DeviceEntry,
+) -> None:
+    """Profile the websocket."""
+    if protect.bootstrap.capture_ws_stats:
+        raise HomeAssistantError("Profile already in progress")
+
+    protect.bootstrap.capture_ws_stats = True
+
+    start_time = time.time()
+    name = device_entry.name_by_user or device_entry.name or device_entry.id
+    nvr_id = name.replace(" ", "_").lower()
+    message_id = f"ufp_ws_profiler_{nvr_id}_{start_time}"
+    hass.components.persistent_notification.async_create(
+        "The WS profile has started. This notification will be updated when it is complete.",
+        title=f"{name}: WS Profile Started",
+        notification_id=message_id,
+    )
+    _LOGGER.info("%s: Start WS Profile for %ss", name, seconds)
+    end = time.monotonic() + seconds
+    while time.monotonic() < end:
+        await asyncio.sleep(10)
+
+    protect.bootstrap.capture_ws_stats = False
+
+    json_data = [s.__dict__ for s in protect.bootstrap.ws_stats]
+    out_path = hass.config.path(f"ufp_ws_profile.{start_time}.json")
+    with open(out_path, "w", encoding="utf8") as outfile:
+        json.dump(json_data, outfile, indent=4)
+
+    stats_buffer = StringIO()
+    print_ws_stat_summary(protect.bootstrap.ws_stats, output=stats_buffer.write)
+    protect.bootstrap.clear_ws_stats()
+    _LOGGER.info("%s: Complete WS Profile:\n%s", name, stats_buffer.getvalue())
+
+    hass.components.persistent_notification.async_create(
+        f"Wrote raw stats to {out_path}\n\n```\n{stats_buffer.getvalue()}\n```",
+        title=f"{name}: WS Profile Completed",
+        notification_id=message_id,
+    )
+
+
+async def generate_sample_data(
+    hass: HomeAssistant,
+    protect: ProtectApiClient,
+    seconds: int,
+    anonymize: bool,
+    device_entry: DeviceEntry,
+) -> None:
+    """Generate sample data from Protect instance."""
+
+    start_time = time.time()
+    name = device_entry.name_by_user or device_entry.name or device_entry.id
+    folder = f"ufp_sample.{start_time}"
+    out_path = hass.config.path(folder)
+    nvr_id = name.replace(" ", "_").lower()
+    message_id = f"ufp_ws_profiler_{nvr_id}_{start_time}"
+    hass.components.persistent_notification.async_create(
+        "The sample data generation has started. This notification will be updated when it is complete.",
+        title=f"{name}: Sample Data Generation Started",
+        notification_id=message_id,
+    )
+
+    _LOGGER.info(
+        "%s: Start Sample Data for %ss (anonymize: %s)", name, seconds, anonymize
+    )
+    await SampleDataGenerator(
+        protect, Path(out_path), anonymize, seconds
+    ).async_generate(close_session=False)
+    shutil.make_archive(out_path, "zip", out_path)
+    shutil.rmtree(out_path)
+    out_path = f"{out_path}.zip"
```

Comment:
This looks like it does I/O in async context.

This should probably move into `pyunifiprotect` instead as well

Reply:
I did not even think of the IO part being blocking. I forgot about that. 

What is the preferred way of doing async IO? aiofiles?

Reply:
Simplest way would be in the executor.

Reply:
await loop.run_in_executor in external code

await hass.async_add_executor_job in hass code